streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamWindowCloseTest.java

```diff
@@ -0,0 +1,150 @@
+        final KStream<Integer, String> stream = builder.stream(LEFT, CONSUMED).join(
+            builder.stream(RIGHT, CONSUMED),
+            MockValueJoiner.TOSTRING_JOINER,
+            WINDOW,
+            StreamJoined.with(Serdes.Integer(), Serdes.String(), Serdes.String())
+        );
+        final MockApiProcessorSupplier<Integer, String, Object, Object> processorSupplier = new MockApiProcessorSupplier<>();
+        stream.process(processorSupplier);
+        return Arguments.of(builder.build(PROPS), processorSupplier);
+    }
+
+    private static Arguments leftJoin() {
+        final StreamsBuilder builder = new StreamsBuilder();
+        final KStream<Integer, String> stream = builder.stream(LEFT, CONSUMED).leftJoin(
+            builder.stream(RIGHT, CONSUMED),
+            MockValueJoiner.TOSTRING_JOINER,
+            WINDOW,
+            StreamJoined.with(Serdes.Integer(), Serdes.String(), Serdes.String())
+        );
+        final MockApiProcessorSupplier<Integer, String, Object, Object> processorSupplier = new MockApiProcessorSupplier<>();
+        stream.process(processorSupplier);
+        return Arguments.of(builder.build(PROPS), processorSupplier);
+    }
+
+    private static Arguments outerJoin() {
+        final StreamsBuilder builder = new StreamsBuilder();
+        final KStream<Integer, String> stream = builder.stream(LEFT, CONSUMED).outerJoin(
+            builder.stream(RIGHT, CONSUMED),
+            MockValueJoiner.TOSTRING_JOINER,
+            WINDOW,
+            StreamJoined.with(Serdes.Integer(), Serdes.String(), Serdes.String())
+        );
+        final MockApiProcessorSupplier<Integer, String, Object, Object> supplier = new MockApiProcessorSupplier<>();
+        stream.process(supplier);
+        return Arguments.of(builder.build(PROPS), supplier);
+    }
+
+    @ParameterizedTest
+    @MethodSource("streams")
+    public void recordsArrivingPostWindowCloseShouldBeDropped(
+        final Topology topology,
+        final MockApiProcessorSupplier<Integer, String, Void, Void> supplier) {
+        try (final TopologyTestDriver driver = new TopologyTestDriver(topology, PROPS)) {
+            final TestInputTopic<Integer, String> left =
+                driver.createInputTopic(KStreamKStreamWindowCloseTest.LEFT, new IntegerSerializer(), new StringSerializer(), Instant.ofEpochMilli(0L), Duration.ZERO);
+            final TestInputTopic<Integer, String> right =
+                driver.createInputTopic(KStreamKStreamWindowCloseTest.RIGHT, new IntegerSerializer(), new StringSerializer(), Instant.ofEpochMilli(0L), Duration.ZERO);
+            final MockApiProcessor<Integer, String, Void, Void> processor = supplier.theCapturedProcessor();
+
+            left.pipeInput(0, "left", 15);
+            right.pipeInput(-1, "bumpTime", 40);
+            assertRecordDropCount(0.0, processor);
+
+            right.pipeInput(0, "closedAt40", 24);
+            assertRecordDropCount(1.0, processor);
+
```

Comment:
If find this test a little bit hard to follow...

We start to open a window `[5;15]` with the first input record. This window would close at `31` so why do we dump to `40`, and test dropping with `ts=24` -- both are totally unrelated to the left input record's window.

Also, it would be good to test that we can join out-of-order record successfully as long as the window is open, and that we don't drop re-mature before we hit the close time, thus I would suggest something like this:
 ```
 // prepare
 - left input at 15 -> open window
 // positive test
 - bump time to 30 (different key) -> window still open
 - right input at 5 -> joins (no need to test with 4, because 4 won't join because of window size anyway...)
 - right input at 25 -> joins 
 // negative test
 - bump time to 31 (different key) -> window closes
 - right input at 5 -> dropped (recorded with metric)
 - right input at 25 -> does not join any longer; window closed (for this case we don't drop and don't record metric)
 // test sharp lower drop bound
  - right input at 6 -> does not join any longer; window closed (for this case we don't drop though and don't record metric)
// cont. with additional sanity check:
 - left input at 16 -> joins with both right input at 6 and right input at 25 -- to verify both records did not get dropped
 ```

Reply:
Ok interesting,  given your test cases, I think misunderstood when we get to drop records.
To check `sharedTimeTracker.streamTime > timeTo + joinGraceMs` ([here](https://github.com/apache/kafka/blob/9e529373df2d157304c03b713b999c2bb1300613/streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java#L132)) seems to be too simplistic.
Because this would already drop the first `right input at 5` beccause (5 + 10 + 5 < 30).
So now I am thinking it needs to include both records in the evaluation.
Something like:
For a matching left and right record only drop the current record if 
`max(thisRecord.ts , otherRecord.ts) + joinAfterMs + grace < sharedTimeTracker.streamTime`.
And that logic would be applied [here](https://github.com/apache/kafka/blob/9e529373df2d157304c03b713b999c2bb1300613/streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java#L154)
However, 

Long story short, I will work on this but wont be able to do so before end of next week... I hope that is ok.

Reply:
Thanks. No rush! We still have time util 3.8 comes around :)

For joins it's a little bit more complicated than for aggregations, but in the end, we need to accept a record as along as a window is open, and window-close time is defined as window-end+grace+1 (for join windows). And for joins, the definition of what a window is, is a little bit more complex (especially when taking "asymmetric" windows into account).

I guess bottom line is, that every input record, does conceptually open a join window over the other input stream, and we need to keep this window open an accept records from the other side until we reach window close time. At the same time, if an out-of-order record comes in and we put it into an existing join window, it might be allowed to open it's own join window any longer, because it's own join window is already too old (this is the weird case that cannot happen for aggregations...)

I did not think about the right/best way how to compute this correctly, but I am sure you'll figure it out :)